### PR TITLE
Feature - System Testing

### PR DESCRIPTION
**Pull request by ArthurCRodrigues**
This pull request adds a complete suite of unit tests for various components in the `systemTests` folder. The tests ensure robust validation of methods, functionalities, and attributes across multiple classes and modules. Below is a summary of the tests added:

### Files Updated:
1. `systemTests/test_Scorer.py`:
   - Tests for the `Scorer` class, including initialization, score calculation, grader addition, feedback generation, and quick build functionality.

2. `systemTests/test_Grader.py`:
   - Tests for the `Grader` class, covering initialization, test result collection, score generation, grading logic, and class method creation.

3. `systemTests/test_SubjectGrader.py`:
   - Tests for the `SubjectGrader` class, including initialization, sub-score generation, unit test scoring, quantitative scoring, and configuration filtering.

4. `systemTests/test_SubTestConfig.py`:
   - Tests for the `SubTestConfig` class, validating configuration loading, quantitative test handling, weight balancing, and string representation.

5. `systemTests/test_TestConfig.py`:
   - Tests for the `TestConfig` class, covering configuration loading, subject handling, weight balancing, and creation of valid/invalid configurations.

6. `systemTests/test_Config.py`:
   - Tests for the `Config` class, including parsing valid/invalid JSON files, loading base/bonus/penalty configurations, and handling quantitative test data.

### Key Features:
- Comprehensive test coverage for all methods and attributes in the respective classes.
- Use of `pytest` fixtures for mock data and instances.
- Validation of edge cases, exceptions, and expected behaviors.
- Mocking external dependencies where necessary.

### Impact:
- Improves code reliability and maintainability.
- Ensures all system components function as expected under various scenarios.
- Facilitates future development by providing a robust testing framework.

### Testing Instructions:
- Run `pytest` in the `systemTests` folder to execute all tests.
- Verify that all tests pass successfully without errors.

### Additional Notes:
- No changes were made to the core functionality; this PR strictly adds unit tests.

### CODE REVIEW

File: systemTests/test_Grader.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    grader_instance.grade.assert_called_once()
+
+def test_grade(grader_instance):
+    """Test the grade method."""
+    grader_instance.passed_tests = ["test1", "test2"]
+    grader_instance.failed_tests = ["test3"]
+
+    score = grader_instance.grade()
+
+    assert score == 2 / 3
```
Comment (gnvr29) [inline systemTests/test_Grader.py:72]: Whigga, doing raw assertions with floats ain't recommended. Try to round it up to a few decimal places to avoid any problems with float precision. -50 aura

File: systemTests/test_SubjectGrader.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    subject_grader_instance.get_unit_tests_score = MagicMock(return_value=40)
+    subject_grader_instance.get_quantitative_score = MagicMock(return_value=20)
+    subject_grader_instance.generate_sub_score()
+    assert subject_grader_instance.score == 30  # (40 + 20) / 100 * 50
+
+def test_get_unit_tests_score(subject_grader_instance, mock_test_report, mock_sub_config):
+    """Test the get_unit_tests_score method."""
+    mock_sub_config.quantitative_tests_weight = 20
+    score = subject_grader_instance.get_unit_tests_score()
+    assert score == (2 / 3) * 80  # 2 passed out of 3 total, adjusted by unit tests weight
```
Comment (gnvr29) [inline systemTests/test_SubjectGrader.py:56]: Careful with floats. -50 aura

File: systemTests/test_SubTestConfig.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    assert quantitative_tests["test_quantitative1"].weight == 10
+
+def test_balance_weights(sub_test_config_instance, mock_sub_config_data):
+    """Test balancing weights of quantitative tests."""
+    sub_test_config_instance.load(mock_sub_config_data)
+    sub_test_config_instance.quantitative_tests[0].weight = 10
+    sub_test_config_instance.quantitative_tests[1].weight = 20
+    sub_test_config_instance.balance_weights()
+    assert sub_test_config_instance.quantitative_tests[0].weight == 33.33
+    assert sub_test_config_instance.quantitative_tests[1].weight == 66.67
```
Comment (gnvr29) [inline systemTests/test_SubTestConfig.py:79]: Beware of floats. -50

**Review by gnvr29 — CHANGES_REQUESTED** [review]
The Holy and Almighty GabigOps, the paramount branch protector, the utmost merge conflict resolver, the API key concealer, and the fire in the Firewall, whose assertions define the before and after, states the following:

1. **Approximate float assertions:** whenever doing assertions with floats, avoid using the raw values due to imprecisions associated with floats (Java has the option to determine an error margin. I do not possess the knowledge of whether this peasant-like language named after the Python snake also holds this power, much less if it has been made use of in here, but I advise thee to embark on a quest for such arcane skill in case it has not been already bestowed upon you. In such case, inform The Holy and Almighty GabigOps of his slight misconception about thy demi-god-like Python dev deeds and capacity).
2. **Parametrizing**: for tests which require data, such as in the test_Config.py file, consider using parametrization so you can test more edge cases.
3. **Test coverage**: overall thou hast covered a fair amount of edge cases and I would proclaim the tests to be enough. However, it can always get better and more edge cases could be tested. I do believe, though, that these will show up and prove themselves necessary throughout the project and it isn't worth wasting that much time pondering about them at least for the next fortnight.

Thou shalt reflect upon the given commandments above and I grant thee the liberty to decide whether the advised changes shall be applied or to keep a code unworthy of my reviews and presence. 

(Jokes apart, I suggest addressing the float topic, as well as considering parametrizing for now. Do as u wish dawg).

The Holy and Almighty GabigOps, the paramount branch protector, the utmost merge conflict resolver, the API key concealer, and the fire in the Firewall, whose assertions define the before and after OUT.
